src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -3655,5 +3682,998 @@ private static ImmutableArray<Symbol> GetProperties(CSharpCompilation comp, stri
+            comp.VerifyDiagnostics(
+                // (2,1): warning CS1717: Assignment made to same variable; did you mean to assign something else?
+                // record C
+                Diagnostic(ErrorCode.WRN_AssignmentToSelf, @"record C
+{
+    private static int _nextId = 0;
+    private int _id;
+    public C() { _id = _nextId++; }
+}").WithLocation(2, 1));
+            var verifier = CompileAndVerify(comp, expectedOutput:
+@"True
+False
+True");
+            verifier.VerifyIL("C.Equals(C)",
+@"{
+  // Code size       42 (0x2a)
+  .maxstack  3
+  IL_0000:  ldarg.1
+  IL_0001:  brfalse.s  IL_0028
+  IL_0003:  ldarg.0
+  IL_0004:  callvirt   ""System.Type C.EqualityContract.get""
+  IL_0009:  ldarg.1
+  IL_000a:  callvirt   ""System.Type C.EqualityContract.get""
+  IL_000f:  bne.un.s   IL_0028
+  IL_0011:  call       ""System.Collections.Generic.EqualityComparer<int> System.Collections.Generic.EqualityComparer<int>.Default.get""
+  IL_0016:  ldarg.0
+  IL_0017:  ldfld      ""int C._id""
+  IL_001c:  ldarg.1
+  IL_001d:  ldfld      ""int C._id""
+  IL_0022:  callvirt   ""bool System.Collections.Generic.EqualityComparer<int>.Equals(int, int)""
+  IL_0027:  ret
+  IL_0028:  ldc.i4.0
+  IL_0029:  ret
+}");
+        }
+
+        [Fact]
+        public void Equality_04()
+        {
+            var source =
+@"using static System.Console;
+record A;
+record B1(int P) : A
+{
+    internal B1() : this(0) { } // PROTOTYPE: Remove
+    internal int P { get; set; } // PROTOTYPE: Remove
+}
+record B2(int P) : A
+{
+    internal B2() : this(0) { } // PROTOTYPE: Remove
+    internal int P { get; set; } // PROTOTYPE: Remove
+}
+class Program
+{
+    static B1 NewB1(int p) => new B1 { P = p }; // PROTOTYPE: Replace with new B1(P)
+    static B2 NewB2(int p) => new B2 { P = p }; // PROTOTYPE: Replace with new B2(P)
+    static void Main()
+    {
+        WriteLine(new A().Equals(NewB1(1)));
+        WriteLine(NewB1(1).Equals(new A()));
+        WriteLine(NewB1(1).Equals(NewB2(1)));
+        WriteLine(new A().Equals((A)NewB2(1)));
+        WriteLine(((A)NewB2(1)).Equals(new A()));
+        WriteLine(((A)NewB2(1)).Equals(NewB2(1)));
+        WriteLine(NewB2(1).Equals((A)NewB2(1)));
+    }
+}";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview, options: TestOptions.ReleaseExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput:
+@"False
+False
+False
+False
+False
+True
+True");
+            verifier.VerifyIL("A.Equals(A)",
+@"{
+  // Code size       20 (0x14)
+  .maxstack  2
+  IL_0000:  ldarg.1
+  IL_0001:  brfalse.s  IL_0012
+  IL_0003:  ldarg.0
+  IL_0004:  callvirt   ""System.Type A.EqualityContract.get""
+  IL_0009:  ldarg.1
+  IL_000a:  callvirt   ""System.Type A.EqualityContract.get""
+  IL_000f:  ceq
+  IL_0011:  ret
+  IL_0012:  ldc.i4.0
+  IL_0013:  ret
+}");
+            verifier.VerifyIL("B1.Equals(B1)",
```

Comment:
would it be worthwhile to VerifyIL on `B2.Equals(B2)` or perhaps on `B1.Equals(A)`? (It looks like other tests have verified this but only in records with no properties.)